src/core/meshcop/border_agent.cpp

```diff
@@ -151,6 +151,19 @@ void BorderAgent::SetMeshCoPServiceChangedCallback(MeshCoPServiceChangedCallback
     mNotifyMeshCoPServiceChangedTask.Post();
 }
 
+Error BorderAgent::GetMeshCoPServiceTxtData(MeshCoPServiceTxtData *aTxtData)
+{
+    Error             error;
+    MeshCoPTxtEncoder meshCoPTxtEncoder(GetInstance());
+
+    SuccessOrExit(error = meshCoPTxtEncoder.EncodeTxtData());
+    aTxtData->mTxtData = meshCoPTxtEncoder.GetTxtData();
+    aTxtData->mLength  = meshCoPTxtEncoder.GetTxtDataLen();
+
+exit:
+    return error;
```

Comment:
This won't work, as the `meshCoPTxtEncoder` is a local variable (scoped within this function), and it has the buffer we use for TXT data. Then, we return a pointer to this buffer back in `aTxtData->mTxtData`. The buffer will be gone/deallocated once we return and `meshCoPTxtEncoder` goes out of scope.

I suggest changing this so that `MeshCoPServiceTxtData` itself provides the buffer instead?

Reply:
Oh I made a terrible mistake.

Updated to let `MeshCoPServiceTxtData` provide the buffer and also did some changes to let `meshCoPTxtEncoder` fill it directly instead of writing to its own bufffer(to avoid one more copy).

Reply:
No worries at all :)